src/Turbo/README.md

```diff
@@ -0,0 +1,267 @@
+
+The way Turbo works is by relying as much as possible on HTML and the backend server.
+It provides 3 features to achieve this:
+
+### 1. Navigate with Turbo Drive
+
+[Turbo Drive](https://turbo.hotwire.dev/handbook/drive) creates a "single-page app" feel
+for traditional server-side applications. It transforms links clicks and forms submissions
+in AJAX calls in order to re-render the page without any page load. It's an evolution of
+the legacy Turbolinks library.
+
+When using Symfony UX Turbo, Turbo Drive is enabled by default and automatically
+watches for links and forms in your application. Read the
+[Turbo Drive documentation](https://turbo.hotwire.dev/handbook/drive) to learn how to
+configure it.
+
+### 2. Create interactive components with Turbo Frames
+
+[Turbo Frames](https://turbo.hotwire.dev/handbook/frames) allows to decompose pages in parts that
+can be updated interactively. Any links and forms inside a frame are captured, and the frame
+content is automatically updated after receiving a response. Only that particular section of the page
+will be extracted from the response to replace the existing content, regardless of whether the server
+provides a full document, or just a fragment containing an updated version of the requested
+frame.
+
+This behavior allows to build interactive components that mimic the behaviors of single-page
+applications without any added complexity.
+
+For example, using the helpers to build frames:
+
+```twig
+{# messages_list.html.twig #}
+{% extends 'base.html.twig' %}
+
+{% block body %}
+<div id="navigation">Links targeting the entire page</div>
+
+<!-- Start a frame: the ID is the way Turbo will update the frame from AJAX responses -->
+{{ turbo_frame_start('message_1') }}
+
+    <h1>My message title</h1>
+    <p>My message content</p>
+
+    <!--
+    When clicking this link, Turbo will catch the link, make an AJAX call to
+    /messages/1/edit, find the "message_1" frame in the body of the response and
+    replace this frame's content with the new frame's content.
+
+    Here, if /messages/1/edit renders messages_edit.html.twig, the form will be
+    displayed inline in the frame.
+    -->
+    <a href="/messages/1/edit">Edit this message</a>
+
+<!-- End the frame -->
+{{ turbo_frame_end() }}
+
+</body>
+```
+
+```twig
+{# messages_edit.html.twig #}
+<body>
+
+<!-- By using the same ID, this response will update the original content to this new one -->
+{{ turbo_frame_start('message_1') }}
+
+    <!--
+    This form is in a frame: Turbo will catch the form submission, make an AJAX call instead and
+    replace the frame's content with the next frame's content.
+    -->
+    {{ form(message_form) }}
+
+    <!--
+    We can also display another link to go back to the list: the original HTML from frame
+    "message_1" in the list will be displayed again.
+    -->
+    <a href="/messages">Back</a>
+
+<!-- End the frame -->
+{{ turbo_frame_end() }}
```

Comment:
shouldn't this be a tag rather than a couple of `_start` and `_end` functions ?

Reply:
AFAIK, tags do things and functions display things, thus this notation. I do agree tags would probably look nicer though. Not fixed about this.

Reply:
Tags would help in IDE autocompletion (same as for 'apply' and 'endapply'), so it's less error-prone I guess.

Reply:
Do we have any examples of tags that output something?
forms use functions for the same reason.

Reply:
One that is very close to this use-case is NelmioSecurityBundle's CSP tags:

```
{% cspscript %}
<script>
    window.api_key = '{{ api_key }}';
</script>
{% endcspscript %}
```
(the modify the `<script>` tag to add the correct CSP digest)

Reply:
That's a different use case though, as it modifies the content of the tag (a bit like autoescape would). It doesn't display anything by itself.

I think I would be in favor of keeping in functions. It matches the way the rest of Symfony works (esp. forms) and it doesn't have that big of an impact on DX IMO. IDE can usually autocomplete functions as well as tags.

Reply:
Well, these 2 functions are about rendering some HTML opening and closing tags. they are indeed not a kind of twig block

Reply:
What happens if I forget some `turbo_frame_end()` function calls?

Reply:
the HTML tag is not closed, and the browser may auto-close it in a different place than where you expect it to be closed.